### PR TITLE
[core] some non-bucket compaction class

### PR DESCRIPTION
**Pull request by leaves12138**
Submit some classes for non-bucket compaction:

NonBucketTable interface to tag a Table whether it could be non-bucket

CompactionCoordinator and Worker to assign and execute compaction task.

link: https://github.com/apache/incubator-paimon/issues/1224

### CODE REVIEW

**Review by JingsongLi — CHANGES_REQUESTED** [review]
You need to add tests for new classes.

File: paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionCoordinator.java [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+                .computeIfAbsent(partition, PartitionCompactCoordinator::new)
+                .addFiles(
+                        files.stream()
+                                .filter(file -> file.fileSize() < targetFileSize)
+                                .collect(Collectors.toList()));
+    }
+
+    // generate compaction task to the next stage
+    public List<AppendOnlyCompactionTask> compactPlan() {
+        return partitionCompactCoordinators.values().stream()
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionCoordinator.java:62]: judge:
- `PartitionCompactCoordinator ` is empty, remove it.
- If `PartitionCompactCoordinator` is a file, plan 10 times, no other files, just remove it. 
- If `PartitionCompactCoordinator` has some file, plan 5 times, no other files, just compact them.

File: paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyNonBucketCompaction.java [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.append;
+
+import org.apache.paimon.table.AppendOnlyFileStoreTable;
+
+/** Compaction for {@link AppendOnlyFileStoreTable}. */
+public class AppendOnlyNonBucketCompaction {
```
[the remark this answers is not in the document]
Comment (leaves12138) [reply paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyNonBucketCompaction.java:24]: agree, removed

File: paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionWorker.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+/** {@link AppendOnlyFileStoreTable} compact worker. */
+public class AppendOnlyTableCompactionWorker {
+    private final List<AppendOnlyCompactionTask> tasks = new ArrayList<>();
+    private final List<CommitMessage> result = new ArrayList<>();
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionWorker.java:35]: should be a local field

File: paimon-core/src/main/java/org/apache/paimon/table/NonBucketTable.java [hunk 4]
```diff
@@ -0,0 +1,26 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.paimon.table;
+
+/** Table bucket could be negative. */
+public interface NonBucketTable {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/table/NonBucketTable.java:22]: remove this class.

File: paimon-core/src/main/java/org/apache/paimon/utils/BucketUtils.java [hunk 5]
```diff
@@ -0,0 +1,58 @@
...
+
+package org.apache.paimon.utils;
+
+import org.apache.paimon.data.BinaryRow;
+import org.apache.paimon.table.NonBucketTable;
+import org.apache.paimon.table.Table;
+import org.apache.paimon.table.sink.KeyAndBucketExtractor;
+
+/** Processor for bucket, including calculate bucket num, handle non bucket, etc. */
+public class BucketUtils {
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/utils/BucketUtils.java:27]: remove this class

File: paimon-core/src/test/java/org/apache/paimon/append/AppendOnlyTableCompactionWorkerTest.java [hunk 6]
```diff
@@ -0,0 +1,172 @@
...
+
+import java.nio.file.Path;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Random;
+import java.util.UUID;
+
+/** Tests for {@link AppendOnlyTableCompactionWorker}. */
+public class AppendOnlyTableCompactionWorkerTest {
+    @TempDir Path tempDir;
```
Comment (JingsongLi) [inline paimon-core/src/test/java/org/apache/paimon/append/AppendOnlyTableCompactionWorkerTest.java:53]: left one empty line

File: paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionCoordinator.java [hunk 7]
```diff
@@ -0,0 +1,220 @@
...
+
+        private List<AppendOnlyCompactionTask> pickCompact() {
+            List<List<DataFileMeta>> waitCompact = agePack();
+            return waitCompact.stream()
+                    .map(files -> new AppendOnlyCompactionTask(partition, files))
+                    .collect(Collectors.toList());
+        }
+
+        public void addFiles(List<DataFileMeta> dataFileMetas) {
+            toCompact.addAll(dataFileMetas);
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionCoordinator.java:146]: here should age = 0;
You should add tests to cover this.

File: paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionCoordinator.java [hunk 8]
```diff
@@ -0,0 +1,217 @@
...
+        // scan files in snapshot
+        scan();
+
+        // do plan compact tasks
+        return compactPlan();
+    }
+
+    @VisibleForTesting
+    void scan() {
+        scan.plan()
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionCoordinator.java:70]: while (true) {
    if (plan is null) break;
}

File: paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionCoordinator.java [hunk 9]
```diff
@@ -0,0 +1,224 @@
...
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.stream.Collectors;
+
+/** {@link AppendOnlyFileStoreTable} compact coordinator. */
```
Comment (JingsongLi) [inline paimon-core/src/main/java/org/apache/paimon/append/AppendOnlyTableCompactionCoordinator.java:39]: Document about: Fragmented files and overwrite snapshot.

**Review by JingsongLi — APPROVED** [review]
+1

**Pull request merged**